src/Symfony/Component/JsonEncoder/ValueTransformer/DateTimeValueTransformer.php

```diff
@@ -9,61 +9,81 @@
  * file that was distributed with this source code.
  */
 
-namespace Symfony\Component\JsonEncoder\Decode\Denormalizer;
+namespace Symfony\Component\JsonEncoder\ValueTransformer;
 
 use Symfony\Component\JsonEncoder\Exception\InvalidArgumentException;
 use Symfony\Component\TypeInfo\Type;
-use Symfony\Component\TypeInfo\Type\BuiltinType;
 use Symfony\Component\TypeInfo\TypeIdentifier;
+use Symfony\Component\TypeInfo\Type\BuiltinType;
 
 /**
- * Casts string to DateTimeInterface.
+ * Transforms DateTimeInterface to string and the other way around during the encoding/decoding process.
  *
  * @author Mathias Arlaud <mathias.arlaud@gmail.com>
  *
- * @internal
+ * @experimental
  */
-final class DateTimeDenormalizer implements DenormalizerInterface
+final class DateTimeValueTransformer implements ValueTransformerInterface
 {
     public const FORMAT_KEY = 'date_time_format';
 
+    /**
+     * @internal
+     */
     public function __construct(
         private bool $immutable,
     ) {
     }
 
-    public function denormalize(mixed $normalized, array $options = []): \DateTime|\DateTimeImmutable
+    /**
+     * @internal
+     */
+    public function toJsonValue(mixed $phpValue, array $options = []): string
+    {
+        if (!$phpValue instanceof \DateTimeInterface) {
+            throw new InvalidArgumentException('The PHP value must implement the "\DateTimeInterface".');
+        }
+
+        return $phpValue->format($options[self::FORMAT_KEY] ?? \DateTimeInterface::RFC3339);
+    }
+
+    /**
+     * @internal
+     */
+    public function toPhpValue(mixed $jsonValue, array $options = []): \DateTime|\DateTimeImmutable
```

Comment:
the return type could be documented using a generic annotation maybe?
Actually, my preference would be to support only `DateTimeImmutable`. DateTime should die.

Reply:
Agreed. Maybe we can only support `DateTimeImmutable`, and throw a meaningful exception when encountering a `DateTime`?

Reply:
Would work for me.